src/main/kotlin/com/_404/cosmos/springpicnic/scenario/entity/Scenario.kt

```diff
@@ -0,0 +1,19 @@
+package com._404.cosmos.springpicnic.scenario.entity
+
+import javax.persistence.*
+
+@Entity
+@Table(name = "scenario")
+open class Scenario(
+        @Id
+        @GeneratedValue(strategy = GenerationType.AUTO)
+        open var id: Long? = null,
+
+        @Column(name = "game_span")
+        open var gameSpan: Int? = null,
+
+        @Column(name = "load_size")
+        open var loadSize: Int? = null
```

Comment:
JPA 표준에서는 엔티티 클래스가 final이면 안된다고 합니다. kotlin 은 default 가 final 이라서 open 키워드가 필요합니다.  그래서 id ("org.jetbrains.kotlin.plugin.allopen") 이런식으로 자동으로 entity 가 opn 되게 플러그인을 적용한다고 합니다.  Entity 는 그래서 open 을 붙여야하는데.. 다른 column 들에는 open 이 필요없는데 당시에 제가 잘못이해했던 것 같아서 컬럼쪽은 open 을 지워도 될 것 같습니다.

Reply:
오 오키오키!!
[spring-guides/tut-spring-boot-kotlin#persistence-with-jpa](https://github.com/spring-guides/tut-spring-boot-kotlin#persistence-with-jpa)

>In order to make lazy fetching working as expected, entities should be open as described in [KT-28525](https://youtrack.jetbrains.com/issue/KT-28525).
lazy loading 을 위해 open 을 이용하고, entity 생성시마다 open 을 붙이기 싫으니까 플러그인을 쓴다는거같네
나중에 플러그인도 적용해두면 좋을듯!

>Here we don’t use [data classes](https://kotlinlang.org/docs/reference/data-classes.html) with val properties because JPA is not designed to work with immutable classes or the methods generated automatically by data classes. If you are using other Spring Data flavor, most of them are designed to support such constructs so you should use classes like data class User(val login: String, …​) when using Spring Data MongoDB, Spring Data JDBC, etc.
data 클래스를 사용하지 않는 이유는 상속이 안되는거 때문이라기 보다는
JPA는 불변 클래스나, data 클래스에 의해 자동으로 생성된 메소드와 함께 작동하도록 설계되지 않아서 인거같음!

Reply:
https://effectivesquid.tistory.com/entry/Kotlin-JPA-%EC%82%AC%EC%9A%A9%EC%8B%9C-Entity-%EC%A0%95%EC%9D%98
>하지만 data class를 이용하게되면 primary constructor에 명시된 필드를 모두 이용하여 hashCode()를 만들어준다. 이를 피하기 위해 @Id가 붙은 필드(프로퍼티)만 primary constructor에 넣고 나머지 필드(프로퍼티)는 body에 정의하게 되는데 뭔가.. 좀 어색하다..
좋은 글이 있어보여 공유함!